packages/liveblocks-node/src/webhooks.ts

```diff
@@ -0,0 +1,192 @@
+import crypto from "crypto";
+import type { IncomingHttpHeaders } from "http";
+
+export class WebhookHandler {
+  private secretBuffer: Buffer;
+  private static secretPrefix = "whsec_";
+  constructor(
+    /**
+     * The signing secret provided on the dashboard's webhooks page
+     * @example "whsec_wPbvQ+u3VtN2e2tRPDKchQ1tBZ3svaHLm"
+     */
+    secret: string
+  ) {
+    if (!secret) throw new Error("Secret is required");
+    if (typeof secret !== "string") throw new Error("Secret must be a string");
+
+    if (secret.startsWith(WebhookHandler.secretPrefix) === false)
+      throw new Error("Invalid secret, must start with whsec_");
+
+    const secretKey = secret.slice(WebhookHandler.secretPrefix.length);
+    this.secretBuffer = Buffer.from(secretKey, "base64");
+  }
+
+  /**
+   * Verifies a webhook request and returns the event
+   */
+  public verifyRequest(request: WebhookRequest): WebhookEvent {
+    const { webhookId, timestamp, rawSignatures } = this.verifyHeaders(
+      request.headers
+    );
+
+    this.verifyTimestamp(timestamp);
+
+    const signature = this.sign(`${webhookId}.${timestamp}.${request.rawBody}`);
+
+    const expectedSignatures = rawSignatures
+      .split(" ")
+      .map((rawSignature) => {
+        const [, parsedSignature] = rawSignature.split(",");
+        return parsedSignature;
+      })
+      .filter(isNotUndefined);
+
+    if (expectedSignatures.includes(signature) === false)
+      throw new Error(
+        `Invalid signature, expected one of ${expectedSignatures}, got ${signature}`
+      );
+
+    const event = JSON.parse(request.rawBody);
+
+    return event;
```

Comment:
You're effectively assigning an `any`-typed value to a `WebhookEvent` return value here implicitly. In this case that's safe because we just verified that this value is coming from a Liveblocks origin, but it may be good to make this explicit using:

```suggestion
    const event: unknown = JSON.parse(request.rawBody);

    return event as WebhookEvent;
```

Reply:
Thinking about forward-compatibility, one downside of this return value is that, if we ever start emitting a new, fourth, webhook event type, this would effectively be a breaking change because this current return type promises that there will only ever be max three possible event types.

This means that, strictly speaking, if we ever start emitting a new event type in the future, and we "just return" it, it will not match the TypeScript type we're promising here.

I think the best way to handle this is to inspect the `type` field here before returning it, and if it's a new unknown event type, throw an error saying "unknown event type, please upgrade to a higher version of @liveblocks/node".

That way, the TypeScript is fully correct and the implementation future-proof. We recently ran into an issue just like this in @liveblocks/core, that made it impossible for us to let the server send a new OpCode, because it would crash existing old clients if we did. We had to come up with a dirty hack as a workaround. I don't wish you this future 😉

Reply:
Oh indeed, that's a great catch, I'll ensure that this is done!

We do expect a couple more events to arrive such as `RoomCreatedEvent` and `RoomDeletedEvent`

Thank you for the review, saved a future headache! 😂